pallets/thea-staking/src/lib.rs

```diff
@@ -518,392 +669,526 @@ pub mod pallet {
+					let nominator_stake = nominator_exposure.value;
+					let nominator_part = Perbill::from_rational(nominator_stake, total_stake);
+					// TODO: Check if backing is the same
+					let nominator_payout = nominator_part * relayer_payout;
+					let nominator_payout: u32 = nominator_payout.unique_saturated_into();
+					// Mint Rewards for Nominators
+					T::Currency::deposit_into_existing(&nominator, nominator_payout.into())?;
+				}
 			}
+			Ok(())
 		}
-		Ok(())
-	}
 
-	// Add public immutables and private mutables.
-	pub fn rotate_session() {
-		let session_index = <CurrentIndex<T>>::get();
-		log::trace!(target: "runtime::thea::staking", "rotating session {:?}", session_index);
-		let active_networks = <ActiveNetworks<T>>::get();
-		// map to collect all active relayers to send to session change notifier
-		let mut map: BTreeMap<Network, OnSessionChange<T::AccountId>> = BTreeMap::new();
-		for network in active_networks {
-			log::trace!(target: "runtime::thea::staking", "rotating for relayers of network {:?}", network);
-			let active = Self::move_queued_to_active(network);
-			map.insert(network, active);
-			Self::compute_next_session(network, session_index);
-		}
-		// Increment SessionIndex
-		let new_session_index = session_index.saturating_add(1);
-		<CurrentIndex<T>>::put(new_session_index);
-		T::SessionChangeNotifier::on_new_session(map);
-		Self::deposit_event(Event::NewSessionStarted { index: new_session_index })
-	}
+		// Add public immutables and private mutables.
+		pub fn rotate_session() {
+			let session_index = <CurrentIndex<T>>::get();
+			log::trace!(target: "runtime::thea::staking", "rotating session {:?}", session_index);
+			let active_networks = <ActiveNetworks<T>>::get();
+			for (offender, (percent, reporters)) in <CommitedSlashing<T>>::iter() {
+				if let Some(net) =
+					active_networks.iter().find(|n| <Candidates<T>>::contains_key(n, &offender))
+				{
+					if let Some(to_slash) = <Candidates<T>>::get(net, &offender) {
+						// total amount transfered to treasury - used for reporters award
+						// calculations
+						let mut total_slashed = BalanceOf::<T>::zero();
+						let actual_percent = Percent::from_percent(percent);
+						// slashing relayer's individual stake
+						let amount: BalanceOf<T> = actual_percent * to_slash.individual;
+						// TODO: where to transfer? % > Treasury && % > to reporters
+						if <pallet_balances::Pallet<T> as Currency<_>>::transfer(
+							&offender,
+							&T::TreasuryPalletId::get().into_account_truncating(),
+							amount,
+							ExistenceRequirement::KeepAlive,
+						)
+						.is_ok()
+						{
+							total_slashed = total_slashed.saturating_add(amount);
+							Self::deposit_event(Event::Slashed { offender, amount });
+						}
+						// slash stakers / nominators
+						let nominators_count = to_slash.stakers.len();
+						if nominators_count != 0 {
+							let nominator_stake_avg = Percent::from_rational(1, nominators_count);
```

Comment:
So I found this calculation weird, 
1. We calculate nominators stake avg 
```rust
let nominator_stake_avg = Percent::from_rational(1, nominators_count) 
```
2. We calculate the nominators_portion 
```rust
let nominators_portion = to_slash.total.saturating_sub(to_slash.individual) 
// Total Relayers Stake - Relayers Individual Stake = Nominators Stake 
```
3. We calculate nominator's total amount 
```rust
let nominator_amount_total = nominator_stake_avg * nominator_portion 
```
4. Then we calculate how much each nominator needs to be slashed individually 
```rust
let nominator_amount_individual = actual_percent * nominator_amount_total 
``` 
and then we go on to transfer to the Treasury, So the problem with this is 
1. The calculation for nominators portion assumes that all nominators have equal amount contributed which is most likely not the case, and you can find out individual contributions from the Storage Item `<Stakers<T>>` 
and then you will need to calculate portions for every nominator individually. I'm sorry if I did not communicate this correctly.